mongo/job_store.go

```diff
@@ -173,3 +174,29 @@ func (m *JobStore) GetJob(ctx context.Context, id string) (models.Job, error) {
 func (m *JobStore) Checker(ctx context.Context, state *healthcheck.CheckState) error {
 	return m.healthClient.Checker(ctx, state)
 }
+
+// PutNumberOfTasks updates the number_of_tasks in a particular job, from the collection, specified by its id
+func (m *JobStore) PutNumberOfTasks(ctx context.Context, id string, numTasks int) (err error) {
+	s := m.Session.Copy()
+	defer s.Close()
+	log.Event(ctx, "putting number of tasks", log.Data{"id": id, "numTasks": numTasks})
+
+	updates := make(bson.M)
+	updates["number_of_tasks"] = numTasks
+	updates["last_updated"] = time.Now()
+	err = m.UpdateJob(updates, s, id)
+
+	return err
+}
+
+// UpdateJob updates a particular job with the values passed in through the 'updates' input parameter
+func (m *JobStore) UpdateJob(updates bson.M, s *mgo.Session, id string) error {
+	update := bson.M{"$set": updates}
+	if err := s.DB(m.Database).C(m.Collection).UpdateId(id, update); err != nil {
+		if err == mgo.ErrNotFound {
+			return errors.New("the job id could not be found in the jobs collection")
```

Comment:
Question:
Is job not found a state that external packages would need to know about? If so consider making a public package error var for this - It provides the caller with the ability to check for this specific error and act on it if they need to.

Reply:
I'm not sure whether "job not found" is a state that external packages might need to be aware of or not to be honest. My understanding is that this PUT endpoint will only ever get called by a service, rather than by a human, so does that make a difference?

Reply:
Update job can return many different types of error the majority of which would mean returning a 500 status.  There are some cases where the caller will want be able to handle certain errors in a certain way for example:
- Duplicate record
- User not authorised
- Record not found
- etc.

By allowing the caller to identify which error they have it allows them to handle it in the most appropriate way
i.e. return a 404 for error == job not found.

We can do the same by returning a error like this:
```
return errors.New("job not found")
```
**But** if you were to do this then the caller would have to look inside this package at the private internal workings to know what error text is used for each case. This is bad because by writing code with knowledge of the internal workings of another package you are now tightly coupled to it. If that error text were to change in future your code would fail and you would have no warning of it. You might not own that code either so it could be updated or changed and you would have no control over it.

By making public error vars we allow the caller to make the same checks on an error returned without having to look at the inner details of the package. Also by making them public they cannot be changed without breaking backwards compatibility - essentially we are making a promise on how this package will work and we have to adhere to it.

Its fairly common pattern in Go the `mgo` package you're using here does it and the [io package](https://golang.org/pkg/io/#PipeReader.Read) below uses public vars to indicate specific error scenarios

Reply:
Thanks David, I think I get it now, I've changed it to do this.